### PR TITLE
cmd/syncthing: Make API serve loop more robust (fixes #3136)

### LINKED ISSUE #3136: Toggling HTTP/HTTPS breaks GUI/API
Toggling the setting and saving the config in the GUI results in the log line:

```
[SYNO4] 20:46:01 WARNING: API: accept tcp 127.0.0.1:8385: use of closed network connection
```

And then there is nothing listening on the GUI port any more.

v0.13.1

### PR DESCRIPTION
**Pull request by calmh**
This is an alternate to canton7's PR. It sacrifices the ability to return an error when creating the service for being more persistent in keeping it running. I think this is worth it.

Tested by repeatedly toggling all GUI things and it still working...

### CODE REVIEW

File: cmd/syncthing/gui.go [hunk 1]
```diff
@@ -350,31 +349,31 @@ func (s *apiService) Serve() {
...
+		// Shutting down permanently
+		l.Debugln("shutting down (stop)")
 	case <-s.configChanged:
-	case <-time.After(time.Second):
-		l.Warnln("API:", err)
+		// Soft restart due to configuration change
+		l.Debugln("restarting (config changed)")
+	case <-serveError:
+		// Restart due to listen/serve failure
+		l.Debugln("restarting:", err)
```
Comment (AudriusButkevicius) [inline cmd/syncthing/gui.go]: Why not log the error here?

**Comment by canton7** [comment]
LGTM (apart from the whole not-building thing).

Just try changing the listen port in the GUI, then trying to connect to both old and new ports. In some of my testing, I was getting a case where I was still able to connect to the old port, even after the listener had (supposedly) been closed, so I'd be interested to see if it's affecting this as well...

File: cmd/syncthing/gui.go (hunk 1 above)
Comment (calmh) [inline cmd/syncthing/gui.go]: Yeah this should never happen so a warnln is probably appropriate

**Comment by calmh** [comment]
I made failure on first startup fatal again, as it was before, and rejiggered the tests a little.

I can't reproduce what you saw canton7 with this at least. However when I hit reload the first time in my browser after changing port, it does display the GUI (but it fails to load any data). That's just my browser doing something braindead with the cached resources though.

**Comment by canton7** [comment]
> That's just my browser doing something braindead with the cached resources though.

What's it doing, out of curiosity?

**Comment by calmh** [comment]
Not reloading the index.html at the very least, not sure what else.

**Comment by AudriusButkevicius** [comment]
@st-review lgtm

**Comment by st-review** [comment]
@AudriusButkevicius: Noted! Need another LGTM or explicit merge command.

**Comment by canton7** [comment]
Mind opening another issue with some details? #3221 *should* have sorted out any caching issues, so it would be good to see if it hasn't...

**Comment by calmh** [comment]
As far as I'm concerned it's just browsers being browsers and nothing related to this

**Comment by calmh** [comment]
@canton7 I think we should drop the noCacheMiddleware much later in the chain so it applies for all requests, and it sets more anti-cache headers. But after this is merged.

**Comment by canton7** [comment]
Yeah, agreed it's not related to this -  it's just that browsers being annoying browsers came up. 

@st-review lgtm too

**Comment by AudriusButkevicius** [comment]
Probably needs to be as the first message with no extra content

**Comment by canton7** [comment]
@st-review lgtm

**Comment by st-review** [comment]
:ok_hand: Merged as 5aacfd16395efc6a375e21113ded6cb3b0d5783e. Thanks, @calmh!

**Pull request closed without merging**